File: apps/windows/SJTUCanvasDownloader/Services/EngineLocator.cs
namespace CanvasDownloader.Services;

/// <summary>
/// Finds the engine binary. Installed layout: <c>engine\sjtu-canvas-engine.exe</c>
/// next to SJTUCanvasDownloader.exe. During development the repository build
/// output is used, or SJTU_CANVAS_ENGINE.
/// </summary>
public static class EngineLocator
{
    private const string EngineFile = "sjtu-canvas-engine.exe";

    public static string Locate()
    {
        var baseDir = AppContext.BaseDirectory;
        // A debug app uses the debug engine (only it has the test mode).
#if DEBUG
        string[] profiles = ["debug", "release"];
#else
        string[] profiles = ["release", "debug"];
#endif
        var engine = FirstExisting(
            Environment.GetEnvironmentVariable("SJTU_CANVAS_ENGINE"),
            Path.Combine(baseDir, "engine", EngineFile),
            Path.Combine(baseDir, EngineFile),
            RepoPath(baseDir, "engine", "target", profiles[0], EngineFile),
            RepoPath(baseDir, "engine", "target", profiles[1], EngineFile));
        return engine ?? throw new FileNotFoundException($"找不到下载引擎 {EngineFile}，请重新安装应用。");
    }

    private static string? FirstExisting(params string?[] candidates) =>
        candidates.FirstOrDefault(path => !string.IsNullOrEmpty(path) && File.Exists(path));

    /// <summary>Walks up from the app folder to a repository checkout.</summary>
    private static string? RepoPath(string start, params string[] parts)
    {
        var directory = new DirectoryInfo(start);
        while (directory is not null)
        {
            if (File.Exists(Path.Combine(directory.FullName, "engine", "Cargo.toml")))
            {
                return Path.Combine([directory.FullName, .. parts]);
            }
            directory = directory.Parent;
        }
        return null;
    }
}

File: apps/macos/Sources/SJTUCanvasDownloader/Services/EngineLocator.swift
import Foundation

enum EngineLocator {
    /// SJTU Canvas Downloader.app/Contents/MacOS/sjtu-canvas-engine. During
    /// development (`swift run`) the engine comes from engine/target — the
    /// debug build first for a debug app, since only it has the test mode.
    /// SJTU_CANVAS_ENGINE overrides both.
    static func locate() throws -> URL {
        let environment = ProcessInfo.processInfo.environment
        var candidates: [URL] = []
        if let path = environment["SJTU_CANVAS_ENGINE"], !path.isEmpty {
            candidates.append(URL(fileURLWithPath: path))
        }
        if let bundled = Bundle.main.url(forAuxiliaryExecutable: "sjtu-canvas-engine") {
            candidates.append(bundled)
        }
        if let repository = repositoryRoot() {
            #if DEBUG
            let profiles = ["debug", "release"]
            #else
            let profiles = ["release", "debug"]
            #endif
            for profile in profiles {
                candidates.append(repository.appendingPathComponent("engine/target/\(profile)/sjtu-canvas-engine"))
            }
        }
        guard let engine = candidates.first(where: { FileManager.default.isExecutableFile(atPath: $0.path) }) else {
            throw EngineError(code: "engine_missing", message: "找不到下载引擎 sjtu-canvas-engine，请重新安装应用。")
        }
        return engine
    }

    /// The source checkout when running from apps/macos/.build.
    private static func repositoryRoot() -> URL? {
        var directory = Bundle.main.executableURL?.deletingLastPathComponent()
        for _ in 0..<8 {
            guard let current = directory else { return nil }
            if FileManager.default.fileExists(atPath: current.appendingPathComponent("engine/Cargo.toml").path) {
                return current
            }
            directory = current.deletingLastPathComponent()
        }
        return nil
    }
}

/// Where the app keeps its data (download list, settings, the encrypted
/// login): ~/Library/Application Support/SJTU Canvas Downloader, or
/// SJTU_CANVAS_DATA_DIR for development and tests.
enum DataLocation {
    static var current: URL {
        if let path = ProcessInfo.processInfo.environment["SJTU_CANVAS_DATA_DIR"], !path.isEmpty {
            return URL(fileURLWithPath: path, isDirectory: true)
        }
        let support = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        return support.appendingPathComponent("SJTU Canvas Downloader", isDirectory: true)
    }

    /// The user's Downloads folder; new downloads go to "SJTU Canvas" inside it.
    static var downloadsFolder: URL? {
        if let path = ProcessInfo.processInfo.environment["SJTU_CANVAS_DOWNLOADS_FOLDER"], !path.isEmpty {
            return URL(fileURLWithPath: path, isDirectory: true)
        }
        return FileManager.default.urls(for: .downloadsDirectory, in: .userDomainMask).first
    }
}
